source/extensions/matching/input_matchers/metadata/matcher.cc

```diff
@@ -0,0 +1,32 @@
+#include "source/extensions/matching/input_matchers/metadata/matcher.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Matching {
+namespace InputMatchers {
+namespace Metadata {
+
+Matcher::Matcher(Envoy::Matchers::ValueMatcherConstSharedPtr value_matcher, bool invert)
+    : value_matcher_(value_matcher), invert_(invert) {}
+
+bool Matcher::match(const Envoy::Matcher::MatchingDataType& input) {
+  if (absl::holds_alternative<absl::monostate>(input)) {
+    return false;
+  }
+
+  if (auto* ptr = absl::get_if<std::shared_ptr<::Envoy::Matcher::CustomMatchData>>(&input);
+      ptr != nullptr) {
+    Matching::Http::MetadataInput::MetadataMatchData* match_data =
+        dynamic_cast<Matching::Http::MetadataInput::MetadataMatchData*>((*ptr).get());
```

Comment:
`(*ptr).get()` is odd -> doesn't ptr work?

Reply:
with only `ptr`, I get
> error: 'std::shared_ptr<Envoy::Matcher::CustomMatchData>' is not polymorphic

Changed to `ptr->get()` though